extensions/migrate.md

```diff
@@ -0,0 +1,321 @@
+    has resumed its session on **NEW**, changes to client state can only come
+    from other servers since **OLD** is no longer processing messages from the
+    client. This fact can make it easier to reason about the migration process.
+
+3. **OLD** sends any data it has about the client to **NEW** that **NEW** does
+   not already have, such as client capabilities. During this time, **NEW** is
+   receiving messages intended for the client. These messages should be
+   processed by **NEW** as if the client was connected, and any output should be
+   buffered.
+
+  * Note that **NEW** already has the vast majority of information it needs
+    about the client, such as nickname, user modes, services account name,
+    channel membership and status, away status, and so on. Very little data is
+    truly local only to **OLD**.
+
+  * None of the data transferred during this step or received from other servers
+    should affect buffering. If this is not the case, a smarter buffering
+    technique is necessary so that formatting can be delayed.
+
+4. Once all the data has has been transferred, **OLD** notifies the client with
+   `MIGRATE PROCEED` and can close the connection.
+
+5. After some time, the client will connect to **NEW** and request to resume the
+   session. After the client sends `MIGRATE END`, **NEW** drains its buffer.
+
+If any internal (non-leaf) servers do not understand the "flip" message, then
+migration is not possible. A leaf only needs to identify if a client is local or
+not, while internal servers take part in routing messages through the tree.
+Since leaves can become internal servers through regular routing rules, then all
+servers must be able to process the "flip" message for migration to be possible.
+
+# Protocol details
+
+## REQUEST
+
+Clients can send `MIGRATE REQUEST` to suggest a migration.
+
+    MIGRATE REQUEST cactus.desert :Free pizza
+
+The parameters are
+
+1. `REQUEST`, indicating the `MIGRATE REQUEST` operation
+
+2. The network-internal name of the server the client wishes to be migrated to.
+  Depending on the IRC network's configuration, this could be different from the
+  address the client ultimately uses to connect to the new server.
+
+3. An optional reason message.
+
+Servers receiving this message can choose to migrate the client to the requested
+server, a different server, or to reject the migration. There is no protocol for
+acknowledging the request: the server simply starts a migration if the request
+is accepted.
+
+## START
+
+The server sends a `MIGRATE START` message to the client to indicate that a
+migration is starting.
+
+    :saguaro.desert MIGRATE START cactus.desert-network.com +6697 5AEAD4D0
+    9DD9F7F5 :Server going down for maintenance
+
+The parameters are
+
+1. `START`, identifying the command
+
+2. The client-reachable address of the server that the client will eventually be
+   migrated to. This can be either a domain name or an IP address, but should be
```

Comment:
IP address in text format, I assume, not DCC format?

Reply:
Yes, text format. I'll clarify.